DAOS-7177 control: report nvme result from storage prepare

tanabarr opened the pull request:
Update dmg storage prepare command output to display results of SCM and
NVMe prepare operations in a way that will illustrate which actions were
performed in the cases that either one or the other failed and if only
one or the other was requested. If a particular host fails NVMe prepare
but succeeds SCM prepare for example, a host error will be displayed for
the NVMe failure and the coalesced result table will contain a separate
entry for the failed host indicating SCM success but no NVMe result.

Update hashstructure package to v2 in go.mod and update vendor directory.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

tanabarr committed:
DAOS-7177 control: report nvme result from storage prepare

Update dmg storage prepare command output to display results of SCM and
NVMe prepare operations in a way that will illustrate which actions were
performed in the cases that either one or the other failed and if only
one or the other was requested. If a particular host fails NVMe prepare
but succeeds SCM prepare for example, a host error will be displayed for
the NVMe failure and the coalesced result table will contain a separate
entry for the failed host indicating SCM success but no NVMe result.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/1/execution/node/75/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

kjacque commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -78,31 +112,53 @@ func newPrepareScmResp(inResp *scm.PrepareResponse, inErr error) (*ctlpb.Prepare
+
+	if req.Nvme == nil {
+		// nil resp indicate no action performed
+		resp.Nvme = nil
```
Not necessary - should already be nil

kjacque commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -78,31 +112,53 @@ func newPrepareScmResp(inResp *scm.PrepareResponse, inErr error) (*ctlpb.Prepare
+
+	if req.Scm == nil {
+		// nil resp indicate no action performed
+		resp.Scm = nil
```
Same note as above

kjacque reviewed (commented):
Small comments. It wasn't clear to me - wasn't the initial PR this was based on landed? Maybe I'm just confused but it looked like there were repeats of some of those commits showing up in the diff even though you changed the base.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

tanabarr commented:
> Small comments. It wasn't clear to me - wasn't the initial PR this was based on landed? Maybe I'm just confused but it looked like there were repeats of some of those commits showing up in the diff even though you changed the base.

Right I had to merge with master after the parent Branch landed, Ended up as quite a big change but the result is more reliable and the results differentiate between independently failing NVMe and SCM results on the same host during storage prepare, see the pretty printer tests for examples

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/4/execution/node/459/log

daosbuild1 commented:
Test stage Build on Ubuntu 20.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/4/execution/node/396/log

daosbuild1 commented:
Test stage Build on CentOS 7 release completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/4/execution/node/454/log

daosbuild1 commented:
Test stage Build on CentOS 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/4/execution/node/465/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/4/execution/node/342/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/4/execution/node/358/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/4/execution/node/466/log

tanabarr commented:
example: https://github.com/daos-stack/daos/pull/5350/files#diff-c5afc62b3182cbf5244a7aff179ae26e42f4c7975ac312746217a55c9ee78d07R857

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/5/execution/node/80/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/6/execution/node/80/log

tanabarr commented:
Check patch failure identified here: https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-5350/6/pipeline/43#step-80-log-48 concerns a spelling mistake in a vendored dependency "hashstructure" and therefore should not be fixed here.

tanabarr commented:
I initially thought about using the request flags to filter out the components of the response to print, but I don't want to rely on the correctness of the flags sent to the server (i.e. if some reason the server prepares SCM then I want to know about it, regardless of what was passed to it in the request, and yes the argument can be made that we should be able to operate with the assumption that the behaviour will be correct and the server will only ever perform what it is requested to do, but I would prefer to display the results that are returned Rather than mask based on what was requested)

in the specific case of prepare and scm/nvme results, if we have multiple hosts and on one of those hosts SCM succeeds but NVMe fails (whereas on the rest of the hosts both succeed), we will have a host failure that can be printed in a separate table but to correlate the NVMe failure with the host storage part of the response we need to return a differentiated value for NvmeDevices (otherwise the aggregated results will not reflect the failure),hence needing the ability to differentiate between a nil and empty slice when hashing.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/7/execution/node/80/log

tanabarr commented:
Example output storage prepare with this PR:
```
bash-4.2$ dmg storage prepare -n
Storage Prepare:
  Hosts          SCM Namespaces Reboot Required NVMe Result
  -----          -------------- --------------- -----------
  wolf-[130-133] N/A            false           OK
bash-4.2$ dmg storage prepare
Memory allocation goals for SCM will be changed and namespaces modified, this will be a destructive operation. Please ensure namespaces are unmounted and locally attached SCM & NVMe devices are not in use. Please be patient as it may take several minutes and subsequent reboot maybe required.
Are you sure you want to continue? (yes/no)
yes
Storage Prepare:
  Hosts          SCM Namespaces        Reboot Required NVMe Result
  -----          --------------        --------------- -----------
  wolf-[130-133] 6.4 TB (2 namespaces) false           OK
bash-4.2$ dmg storage prepare --reset
Memory allocation goals for SCM will be changed and namespaces modified, this will be a destructive operation. Please ensure namespaces are unmounted and locally attached SCM & NVMe devices are not in use. Please be patient as it may take several minutes and subsequent reboot maybe required.
Are you sure you want to continue? (yes/no)
yes
Storage Prepare:
  Hosts          SCM Namespaces     Reboot Required NVMe Result
  -----          --------------     --------------- -----------
  wolf-[130-133] 0 B (0 namespaces) true            OK
```

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mjmac reviewed (changes requested):
I have been going around and around on this one. I understand and agree with the intent, which is to accurately convey to the user what exactly has happened on the servers in response to a `StoragePrepare()` RPC. I've also spent quite a bit of time trying to figure out how I can help you achieve your goal here without needing so many significant changes to the codebase.

What I don't like about this implementation is that it changes the `HostStorage` struct from a relatively straightforward representation of a particular host storage configuration into an even more complex and context-sensitive representation of the results of the unified `StoragePrepare` operation. As a result, there is a rather wide ripple of changes out into unrelated code. These kinds of changes are the opposite of what we're trying to do at this point.

Instead of adding all of this new complexity to an already-complex type and potentially narrowing its utility outside of a limited set of use cases, I think it would be a better investment to look at simplifying the storage prepare workflow by using what we have in better ways.

For example, the current way we handle `dmg storage prepare` is to send a single `StoragePrepare()` RPC to each server in the hostlist, and then we pull apart the protobuf responses to convert them into `HostStorage` structs which allow us to group like host configurations so that we can meaningfully display results for large numbers of servers with homogeneous storage configurations. What you're attempting to do with this PR is to handle the use cases where an admin wants to prepare only SCM or only NVMe, and you want to be sure to accurately represent what actually happened on each server.

I propose that we should make the "prepare SCM" and "prepare NVMe" operations distinct at the API/RPC level. We don't need to change the dmg command surface, but behind the scenes we would be calling two separate RPCs on each server. In the case where the admin doesn't specify that one or the other should be prepared, then we just call `StoragePrepareScm()` first, and if it succeeded we then call `StoragePrepareNvme()`. We can then display the results from each set of operations in a coherent way with confidence that we're accurately reflecting the current state of storage on the affected servers.

This change would simplify the logic on the server side, which has to do a lot of accounting to accurately convey the results of a combined operation, it would simplify the display helpers, and it would simplify a lot of the testing as well. I recognize that this represents quite a significant change in direction from where you took this PR, but I can't support landing a disruptive change like this knowing that much would just have to be undone later anyhow.

tanabarr commented:
> I have been going around and around on this one. I understand and agree with the intent, which is to accurately convey to the user what exactly has happened on the servers in response to a `StoragePrepare()` RPC. I've also spent quite a bit of time trying to figure out how I can help you achieve your goal here without needing so many significant changes to the codebase.
> 
> What I don't like about this implementation is that it changes the `HostStorage` struct from a relatively straightforward representation of a particular host storage configuration into an even more complex and context-sensitive representation of the results of the unified `StoragePrepare` operation. As a result, there is a rather wide ripple of changes out into unrelated code. These kinds of changes are the opposite of what we're trying to do at this point.
> 
> Instead of adding all of this new complexity to an already-complex type and potentially narrowing its utility outside of a limited set of use cases, I think it would be a better investment to look at simplifying the storage prepare workflow by using what we have in better ways.
> 
> For example, the current way we handle `dmg storage prepare` is to send a single `StoragePrepare()` RPC to each server in the hostlist, and then we pull apart the protobuf responses to convert them into `HostStorage` structs which allow us to group like host configurations so that we can meaningfully display results for large numbers of servers with homogeneous storage configurations. What you're attempting to do with this PR is to handle the use cases where an admin wants to prepare only SCM or only NVMe, and you want to be sure to accurately represent what actually happened on each server.
> 
> I propose that we should make the "prepare SCM" and "prepare NVMe" operations distinct at the API/RPC level. We don't need to change the dmg command surface, but behind the scenes we would be calling two separate RPCs on each server. In the case where the admin doesn't specify that one or the other should be prepared, then we just call `StoragePrepareScm()` first, and if it succeeded we then call `StoragePrepareNvme()`. We can then display the results from each set of operations in a coherent way with confidence that we're accurately reflecting the current state of storage on the affected servers.
> 
> This change would simplify the logic on the server side, which has to do a lot of accounting to accurately convey the results of a combined operation, it would simplify the display helpers, and it would simplify a lot of the testing as well. I recognize that this represents quite a significant change in direction from where you took this PR, but I can't support landing a disruptive change like this knowing that much would just have to be undone later anyhow.

To be clear, you are okay with the dmg changes in the PR so the dmg display component of the PR can remain as it is. The main objection is to do with the fallout from changing the HostStorage members to pointers so this needs to be reverted and moving to a 2 stage model will fix this. The sequential scm then nvme sets of responses will result in either host error or an entry in a table hostset for each host request? if one host error is returned from scm stage then what is displayed to the user? presumably the nvme stage doesn't get run in that case so the hosts will have scm prepared but not nvme? does the display show the rest of the results for scm and N/A for nvme in the table? Is my interpretation accurate, please correct if not and help me understand the answers to the questions above about your desired implementation.

tanabarr committed:
DAOS-7219 control: trim ssd details in basic storage scan

When dmg storage scan is run without options, summary should group hosts
with similar storage configurations even if SSD details such as model
and firmware version differ. NVMe namespace capacity differences should
be reflected in summary output. Update hashstructure module to v2.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

tanabarr committed:
increase unit test coverage

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

mjmac commented:
> 
> To be clear, you are okay with the dmg changes in the PR so the dmg display component of the PR can remain as it is. 

Correct.

> The main objection is to do with the fallout from changing the HostStorage members to pointers so this needs to be reverted and moving to a 2 stage model will fix this. The sequential scm then nvme sets of responses will result in either host error or an entry in a table hostset for each host request?

Yes, I think this will make it much easier to create/process the responses and make the results easier for the user to understand.

> if one host error is returned from scm stage then what is displayed to the user? presumably the nvme stage doesn't get run in that case so the hosts will have scm prepared but not nvme?

This was what I suggested in my previous response, but after thinking about this some more I'm not convinced it's correct. My previous suggestion was based on the idea that it doesn't make sense to proceed with preparing NVMe if SCM wasn't prepared. However, I can see a case being made for allowing NVMe prepare to proceed on hosts where SCM prepare succeeded.

Example scenario:
  dmg storage prepare -l host-[00-99]
    (host-42 fails to prepare SCM, but all others succeed)

With my initial suggestion (fail immediately, don't prepare NVMe anywhere), the admin has to go look at host-42 and fix it before trying again, or else has to re-run the command with host-42 excluded from the hostlist. On the hosts that succeeded the first time, the SCM prepare is a no-op, but there is still some cost to calling the RPC again and doing the checking on the server side to determine that it is a no-op.

The alternative, which would preserve current behavior, is that even though we observed that host-42 failed the SCM prepare step, we continue on to the NVMe prepare step while excluding host-42 from that request's hostlist. Then at the end of the operation we show the results to the admin who will then go to fix host-42 and re-run the prepare just for that host.

I think both approaches are valid. The first one is arguably more conservative, in that it allows the admin to detect and address problems more quickly, but it seems less user-friendly for the standard use case. The second approach has the benefit of preserving existing behavior and is probably better in terms of standard user experience, so I think we should take this approach with the idea that we could add a new flag to fail early if that turns into a use case that's important to someone.

> does the display show the rest of the results for scm and N/A for nvme in the table? Is my interpretation accurate, please correct if not and help me understand the answers to the questions above about your desired implementation.

I think the display should basically be the same as what you're trying to do with this PR.

tanabarr commented:
> > To be clear, you are okay with the dmg changes in the PR so the dmg display component of the PR can remain as it is.
> 
> Correct.
> 
> > The main objection is to do with the fallout from changing the HostStorage members to pointers so this needs to be reverted and moving to a 2 stage model will fix this. The sequential scm then nvme sets of responses will result in either host error or an entry in a table hostset for each host request?
> 
> Yes, I think this will make it much easier to create/process the responses and make the results easier for the user to understand.
> 
> > if one host error is returned from scm stage then what is displayed to the user? presumably the nvme stage doesn't get run in that case so the hosts will have scm prepared but not nvme?
> 
> This was what I suggested in my previous response, but after thinking about this some more I'm not convinced it's correct. My previous suggestion was based on the idea that it doesn't make sense to proceed with preparing NVMe if SCM wasn't prepared. However, I can see a case being made for allowing NVMe prepare to proceed on hosts where SCM prepare succeeded.
> 
> Example scenario:
> dmg storage prepare -l host-[00-99]
> (host-42 fails to prepare SCM, but all others succeed)
> 
> With my initial suggestion (fail immediately, don't prepare NVMe anywhere), the admin has to go look at host-42 and fix it before trying again, or else has to re-run the command with host-42 excluded from the hostlist. On the hosts that succeeded the first time, the SCM prepare is a no-op, but there is still some cost to calling the RPC again and doing the checking on the server side to determine that it is a no-op.
> 
> The alternative, which would preserve current behavior, is that even though we observed that host-42 failed the SCM prepare step, we continue on to the NVMe prepare step while excluding host-42 from that request's hostlist. Then at the end of the operation we show the results to the admin who will then go to fix host-42 and re-run the prepare just for that host.
> 
> I think both approaches are valid. The first one is arguably more conservative, in that it allows the admin to detect and address problems more quickly, but it seems less user-friendly for the standard use case. The second approach has the benefit of preserving existing behavior and is probably better in terms of standard user experience, so I think we should take this approach with the idea that we could add a new flag to fail early if that turns into a use case that's important to someone.
> 
> > does the display show the rest of the results for scm and N/A for nvme in the table? Is my interpretation accurate, please correct if not and help me understand the answers to the questions above about your desired implementation.
> 
> I think the display should basically be the same as what you're trying to do with this PR.

So the revised requirements just involve splitting RPCs, running them sequentially, collating the results and returning from the control API the exact same data. Will do.

tanabarr committed:
Merge branch 'master' of github.com:daos-stack/daos into tanabarr/control-scan-nvme-summary

mjmac commented:
> 
> So the revised requirements just involve splitting RPCs, running them sequentially, collating the results and returning from the control API the exact same data. Will do.

I think there is an open question here with regard to whether or not the separate prepare operations should be exposed via the Control API as well. I think for the moment I'm fine with hiding the details behind a single API call unless it means adding a lot of complexity. We could make these changes in a couple of steps (i.e. split things up on the backend first, then rework dmg to call separate APIs and present the results in separate tables).

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

tanabarr commented:
@mjmac I don't think the proposal to `...make the "prepare SCM" and "prepare NVMe" operations distinct at the API/RPC level` will resolve the relevant issue.

The crux of the issue that this PR attempts to resolve is that when building the HostStorageMap that represents groupings of hosts that have responded with identical results, there is no distinction between a host that failed NVMe prepare and one that didn't.

For example even if we run separate RPCs for NVMe we still need to combine the response results and add to the HostStorageMap:
- host1 returns HS with ScmNamespaces: S1 after prepare SCM success
- host1 then returns empty HS from prepare NVMe after success
- host2 returns HS with ScmNamespaces: S1 after prepare SCM success
- host2 then returns empty HS from prepare NVMe after failure with an entry in HEM
In this scenario, both HostStorage structs to be added to the Map will hash to the same value because no hashable HostStorage field is altered to communicate the NVMe failure.

This can be resolved in a number of ways:
- existing solution where fields are changed to references (to slice aliases) and NVMe prepare failure can be represented by nil and success by empty slice (the difference between which results in distinct hash digests)
- postprocessing HostStorageMap to remove/add distinct entries for host entries in the HostErrorsMap
- similar to above but perform postprocessing in pretty printer; alter display of results by extracting entries from HostStorageMap and correlating with HostErrorsMap to reflect the distinction of hosts that have failed NVMe prepare, will require typing/recognition of NVMe specific errors
- add another member to the HostStorage struct which indicates NVMe prepare result explicitly (similar to "reboot required")

Which route would you like me to follow?

tanabarr committed:
Merge branch 'master' of github.com:daos-stack/daos into tanabarr/control-scan-nvme-summary

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/10/execution/node/910/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/11/execution/node/107/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/12/execution/node/105/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/12/execution/node/863/log

kjacque reviewed (commented):
It seems basically sane to me, but one thought I have is how stable should we expect the control API to be at this point? Maybe it would be worthwhile to create a different prepare result map, that could include errors for individual SSD components. The fact that it's hard to shoehorn all we need into the HostStorageMap without some subtlety like this, nil vs. empty, could indicate the use case is better served by a new type.

If you fix the unit test failures, I could approve as-is, but it might be better to step back and see if creating a new type reduces or removes the problems we've gone back and forth on.

tanabarr commented on src/control/lib/control/storage.go:
```diff
@@ -34,19 +34,19 @@ var storageHashOpts = hashstructure.HashOptions{
 	// NvmeDevices contains the set of NVMe controllers (SSDs)
 	// in this configuration.
-	NvmeDevices storage.NvmeControllers `json:"nvme_devices"`
+	NvmeDevices *storage.NvmeControllers `json:"nvme_devices"`
```
yes but hashstructure doesn't distinguish between a typed nil slice and an empty slice when generating a hash digest, so in order to generate distinct keys in the HSM I needed to change the field to a pointer, in which case a nil pointer could be distinguished from a pointer to an empty slice (for failure and success respectively) which would each generate a distinct key for the map when hashed

mjmac replied:
This is the crux of the issue I am having with this PR. Because a third-party library doesn't behave the way we want it to in order to handle this narrow use case, we're making changes all throughout our codebase? What if the next version fixes it, or otherwise changes behavior? Will we have more ripples throughout our code to accommodate that?

Given that the code is open source under a permissive license, I think that we could fork it, fix it, use it and submit the fix upstream. I suspect that they would accept a contribution to fix this because after studying the code I think it's an oversight rather than a deliberate design decision.

Alternatively, we could use this as an opportunity to rethink what we're doing here. I'm still not convinced that we need to perform remote NVMe prepare/reset. The server already does an automatic prepare on startup. This seems to me like we're investing a lot of time and energy on solving a problem with limited application.

tanabarr commented:
> It seems basically sane to me, but one thought I have is how stable should we expect the control API to be at this point? Maybe it would be worthwhile to create a different prepare result map, that could include errors for individual SSD components. The fact that it's hard to shoehorn all we need into the HostStorageMap without some subtlety like this, nil vs. empty, could indicate the use case is better served by a new type.
> 
> If you fix the unit test failures, I could approve as-is, but it might be better to step back and see if creating a new type reduces or removes the problems we've gone back and forth on.

have had a fair amount of discussion with @mjmac on the topic and he is considering his preferred approach

tanabarr committed:
Merge branch 'master' of github.com:daos-stack/daos into tanabarr/control-dmg-prep-results

tanabarr committed:
fix dmg/pretty unit tests

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/13/execution/node/123/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

tanabarr committed:
Merge branch 'master' of github.com:daos-stack/daos into tanabarr/control-dmg-prep-results

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5350/15/execution/node/889/log

tanabarr committed:
Merge branch 'master' of github.com:daos-stack/daos into tanabarr/control-dmg-prep-results

tanabarr committed:
fix unit test failures

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.
